proteuslib/tests/test_chlorination.py

```diff
@@ -0,0 +1,823 @@
+solver = get_solver()
+
+# Start test class
+class TestChlorination():
+    @pytest.fixture(scope="class")
+    def chlorination_obj(self):
+        model = ConcreteModel()
+        model.fs = FlowsheetBlock(default={"dynamic": False})
+        model.fs.thermo_params = GenericParameterBlock(default=thermo_config)
+        model.fs.rxn_params = GenericReactionParameterBlock(
+                default={"property_package": model.fs.thermo_params, **reaction_config})
+        model.fs.unit = EquilibriumReactor(default={
+                "property_package": model.fs.thermo_params,
+                "reaction_package": model.fs.rxn_params,
+                "has_rate_reactions": False,
+                "has_equilibrium_reactions": True,
+                "has_heat_transfer": False,
+                "has_heat_of_reaction": False,
+                "has_pressure_change": False})
+
+        model.fs.unit.inlet.mole_frac_comp[0, "H_+"].fix( 0. )
+        model.fs.unit.inlet.mole_frac_comp[0, "OH_-"].fix( 0. )
+        model.fs.unit.inlet.mole_frac_comp[0, "OCl_-"].fix( 0. )
+        model.fs.unit.inlet.mole_frac_comp[0, "NH4_+"].fix( 0. )
+
+        model.fs.unit.inlet.mole_frac_comp[0, "NH2Cl"].fix( 0. )
+        model.fs.unit.inlet.mole_frac_comp[0, "NHCl2"].fix( 0. )
+        model.fs.unit.inlet.mole_frac_comp[0, "NCl3"].fix( 0. )
+
+        waste_stream_ammonia = 1 #mg/L
+        total_molar_density = 54.8 #mol/L
+        total_ammonia_inlet = waste_stream_ammonia/17000 # mol/L
+        total_molar_density+=total_ammonia_inlet
+
+        # Free Chlorine (mg-Cl2/L) = total_chlorine_inlet (mol/L) * 70,900
+        free_chlorine_added = 15 #mg/L as Cl2
+        total_chlorine_inlet = free_chlorine_added/70900 # mol/L
+        total_molar_density+=total_chlorine_inlet
+
+        model.fs.unit.inlet.mole_frac_comp[0, "NH3"].fix( total_ammonia_inlet/total_molar_density )
+        model.fs.unit.inlet.mole_frac_comp[0, "HOCl"].fix( total_chlorine_inlet/total_molar_density )
+
+        # Perform a summation of all non-H2O molefractions to find the H2O molefraction
+        sum = 0
+        for i in model.fs.unit.inlet.mole_frac_comp:
+            # NOTE: i will be a tuple with format (time, component)
+            if i[1] != "H2O":
+                sum += value(model.fs.unit.inlet.mole_frac_comp[i[0], i[1]])
+
+        model.fs.unit.inlet.mole_frac_comp[0, "H2O"].fix( 1-sum )
+
+        model.fs.unit.inlet.pressure.fix(101325.0)
+        model.fs.unit.inlet.temperature.fix(300.)
+        model.fs.unit.inlet.flow_mol.fix(10)
+
+        return model
+
+    @pytest.mark.unit
+    def test_build_model(self, chlorination_obj):
+        model = chlorination_obj
+
+        assert hasattr(model.fs.thermo_params, 'component_list')
+        assert len(model.fs.thermo_params.component_list) == 10
+        assert 'H2O' in model.fs.thermo_params.component_list
+        assert hasattr(model.fs.thermo_params, 'H2O')
+        assert isinstance(model.fs.thermo_params.H2O, Solvent)

```

Comment:
1) `isinstance` is better than `hasattr` as it both checks that the component exists, and that it has the expected `type`.
2) You need to be careful with both. In these specific cases you are fine, but when dealing with actual properties, you can inadvertently trigger construction.

E.g., consider a case where I do not expect `entr_mol_phase` to exist. If I were to write"

`assert not hasattr(m.entr_mol_phase)`

it will in fact trigger the construction of `entr_mol_phase` (`hasattr` calls `getattr` which builds the component) and thus this test would always fail. Instead, I need to test for a related component that would be built at the same time as `entr_mol_phase`. This can also cause problems if you want to check the value of a component. If the component did not exist before the check, it will be constructed using default values, meaning there is a good chance the value will not be what you expect.